usr/src/uts/common/fs/zfs/vdev.c

```diff
@@ -2489,7 +2489,12 @@ vdev_online(spa_t *spa, uint64_t guid, uint64_t flags, vdev_state_t *newstate)
 		spa_async_request(spa, SPA_ASYNC_CONFIG_UPDATE);
 	}
 
-	if (postevent)
+	if (wasoffline ||
+	    ((oldstate == VDEV_STATE_REMOVED ||
+	    oldstate == VDEV_STATE_CANT_OPEN ||
+	    oldstate == VDEV_STATE_FAULTED) &&
+	    (vd->vdev_state == VDEV_STATE_DEGRADED ||
+	    vd->vdev_state == VDEV_STATE_HEALTHY)))
```

Comment:
Can you try to simplify this further by doing the following:

if (wasoffline || (oldstate < VDEV_STATE_DEGRADED && vd->vdev_state >= VDEV_STATE_DEGRADED))

This should provide the same functionality.